Add `signed(T)`, `unsigned(T)`.

NHDaly opened the pull request:
Converts Types into their corresponding signed/unsigned equivalents, maintaining storage size, but changing signed-ness.

Fixes https://github.com/JuliaLang/julia/issues/30444.

NHDaly committed:
Add `signed(T)`, `unsigned(T)`.

Converts Types into their corresponding signed/unsigned equivalents,
maintaining storage size, but changing signed-ness.

NHDaly committed:
Switch order of `unsigned(x)` and `unsigned(T)` for docstring

NHDaly commented:
(Still needs tests, etc, but just wanted to see what y'all think first! 😄)

JeffreySarnoff reviewed (commented):
I use these extensions often.  For the builtin types,  I favor
```
unsigned(::Type{Int64}) = UInt64
unsigned(::Type{Float32}) = UInt32  # important for low level floating point work
unsigned(x::T) where {T<:Union{Int8..Int128}} = reinterpret(unsigned(T), x)
# etc and cover signed() and e.g. floating(::Type{UInt32}) = Float32 ..
````

timholy reviewed (commented):
I approve of adding this, once it has tests.

@JeffreySarnoff, I don't doubt being able to derive the "plain bits" type corresponding to a floating-point type is useful, but I don't think that operation can be called `unsigned`. Maybe call that `planbitstype` or something.

NHDaly committed:
Restore `[un]signed(x)` impls; use for `signed(T)`

Also, since I restored the original definitions, I also reordered the
methods back to their original places.

NHDaly commented on base/int.jl:
```diff
@@ -159,6 +159,7 @@ julia> signed(unsigned(-2))
 """
 unsigned(x) = convert(Unsigned, x)
+
 signed(x::Unsigned) = reinterpret(typeof(convert(Signed, zero(x))), x)
```
Also, I wanted to ask: Does anyone know why `signed` takes `x::Unsigned`, but `unsigned` takes `x::BitSigned`? I'm _guessing_ it's to prevent `BigInt` from falling under the `reinterpret` implementation. I'm currently just copying that in the `::Type{T}` definitions -- does that seem reasonable?

JeffreySarnoff replied:
yes -- the only difference is the presence/absence of `BigInt`

timholy replied:
Now that you mention it, it is a little weird, because there's nothing about the docstring for `Unsigned` that specifies it needs to be `isprimitivetype`. Nothing stops me from defining

```julia
struct WeirdUnsigned <: Unsigned
    msg::String
    val::UInt16
end
WeirdUnsigned(x::Int) = WeirdUnsigned("missing", UInt16(x))
Base.Signed(x::WeirdUnsigned) = signed(x.val)

julia> signed(WeirdUnsigned(4))
ERROR: bitcast: expected primitive type value for second argument
Stacktrace:
 [1] reinterpret at ./essentials.jl:381 [inlined]
 [2] signed(::WeirdUnsigned) at ./int.jl:162
 [3] top-level scope at none:0
```

Of course this isn't a catastrophe: the error is sensible and one could always provide a more specialized definition. (If it returned a borked answer, that would be a genuine problem.) But of course the same could be said of `Signed`, and `reinterpret(UInt, BigInt(4))` throws the same error.

NHDaly replied:
Yeah, that's exactly the sort of thing I was referring to. Thanks for including an example.

> Of course this isn't a catastrophe: the error is sensible and one could always provide a more specialized definition. (If it returned a borked answer, that would be a genuine problem.) But of course the same could be said of `Signed`, and `reinterpret(UInt, BigInt(4))` throws the same error.

Totally, I agree. But I think that it's the inconsistency with `unsigned()` that's a little weird. I think _either_ we should change this to also accept only `BitUnsigned`, so that you'd get the correct behavior above, _OR_ we should loosen `unsigned()` to accept any kind of `Signed`, and just add an extra overload for `BigInt`, exactly as you'd have to do for `WeirdUnsigned` above. Does that seem reasonable to you?

NHDaly replied:
I've opened https://github.com/JuliaLang/julia/blob/master/base/int.jl#L170 to continue this conversation, since it's not contained to this PR! :)

NHDaly replied:
(oops, i pasted the wrong link before. I've opened https://github.com/JuliaLang/julia/pull/30445.)

NHDaly replied:
(ahhh omg sorry, wrong link again. fixed now. https://github.com/JuliaLang/julia/issues/30473)

NHDaly committed:
Add tests for [un]signed(T). Fix signed(Int)==Int.

Added unit tests, and also added the missing implementations for calling
`signed` on already signed types and same for `unsigned`.

NHDaly commented:
> I approve of adding this, once it has tests.

:) Thanks Tim! I've added Tests. This will probably also need a News change before it could be merged.

> @JeffreySarnoff, I don't doubt being able to derive the "plain bits" type corresponding to a floating-point type is useful, but I don't think that operation can be called `unsigned`. Maybe call that `plainbitstype` or something.

I have to agree with Tim about that. Converting a float to an int seems beyond the scope of `unsigned`/`signed`. Also, I think you could already achieve that with something like `Int(3.0)` / `UInt(3.0)`?

NHDaly commented on base/int.jl:
````diff
@@ -159,7 +156,21 @@ julia> signed(unsigned(-2))
+UInt8
+```
+"""
+unsigned(::Type{T}) where {T<:BitSigned} = typeof(convert(Unsigned, zero(T)))
````
Done. Thanks!

timholy commented on test/int.jl:
```diff
@@ -55,11 +55,27 @@ using Random
-    @test_throws InexactError signed(UInt(-3))
+    @test signed(UInt(0) - 1) === -1
+    @test_throws InexactError signed(UInt(-3))  # Note that UInt() throws, not signed().
+end
```
I don't understand that test either. Does `git blame` shed any light?

NHDaly replied:
Looks like the test was added here: https://github.com/JuliaLang/julia/pull/11962, during `v0.4.0`.

Maybe the behavior was different back then, or maybe it was a mistake? (It was David's _first PR ever_ after all! 😋) I'm happy to either remove the whole line or just leave the comment; whatever you prefer!

NHDaly commented:
Also: is this news-worthy? I'm ambivalent on that, but here's a draft for what i can put there:

```markdown
* There are new method overloads `unsigned(T::Type)` and `signed(T::Type)` that accept a `Type`,
  and return its corresponding signed/unsigned equivalent type ([#30445]).
```

fredrikekre commented on base/int.jl:
```diff
@@ -169,6 +170,29 @@ signed without checking for overflow.
 
+
+"""
+    unsigned(T) -> Unsigned equivalent to T
```
```suggestion
    unsigned(T::Type)
```

fredrikekre commented on base/int.jl:
```diff
@@ -169,6 +170,29 @@ signed without checking for overflow.
+"""
+    unsigned(T) -> Unsigned equivalent to T
+
+Get the corresponding unsigned type for T.
```
```suggestion
Return the corresponding unsigned type for type `T`.
```

fredrikekre commented on base/int.jl:
```diff
@@ -169,6 +170,29 @@ signed without checking for overflow.
+    unsigned(T) -> Unsigned equivalent to T
+
+Get the corresponding unsigned type for T.
+
```
```suggestion

!!! compat "Julia 1.2"
    This method requires Julia 1.2 or later.

```

fredrikekre commented on base/int.jl:
```diff
@@ -169,6 +170,29 @@ signed without checking for overflow.
+unsigned(::Type{T}) where {T<:BitSigned} = typeof(unsigned(zero(T)))
+unsigned(::Type{T}) where {T<:Unsigned} = T
+"""
+    signed(T) -> Signed equivalent to T
```
```suggestion
    signed(T::Type)
```

fredrikekre commented on base/int.jl:
```diff
@@ -169,6 +170,29 @@ signed without checking for overflow.
+"""
+    signed(T) -> Signed equivalent to T
+
+Get the corresponding signed type for T.
```
```suggestion
Return the corresponding signed type for type `T`.
```

fredrikekre commented on base/int.jl:
```diff
@@ -169,6 +170,29 @@ signed without checking for overflow.
+    signed(T) -> Signed equivalent to T
+
+Get the corresponding signed type for T.
+"""
```
````suggestion

!!! compat "Julia 1.2"
    This method requires Julia 1.2 or later.

#Examples
```jldoctest
julia> signed(UInt8)
Int8
```    
"""
````

NHDaly replied:
Thanks. I left off the example to match that `signed(x)` doesn't have examples for some reason. Happy to commit this no problem, though!

fredrikekre committed:
Apply @fredrikekre's suggestions from code review

Fixes docstrings for `unsigned(T::Type)` and `signed(T::Type)`.

Co-Authored-By: NHDaly <NHDaly@gmail.com>

NHDaly committed:
Add NEWS.md entry

NHDaly commented:
(K, i added a NEWS entry based on your thumbsup)

NHDaly committed:
Fix trailing whitespace in docstring

Line 202 had trailing whitespace in the docstring which broke the tests. Fixed.

NHDaly committed:
Remove generated link at bottom of NEWS to fix conflicts

Now that I understand we're not supposed to generate these every time, i've removed the generated links to prevent conflicts.

NHDaly commented:
I reran the tests, and they're passing now. :)

Please take another look when you get a chance! :)

NHDaly commented:
Friendly ping, this should be good to merge. Please take another look?

JeffreySarnoff commented:
I am wondering what benefit there is to avoiding symmetry in these two functional definitions.
To my eye, this is more natural and tends better toward future-proof:
```
unsigned(::Type{T}) where {T<:Unsigned} = T
unsigned(::Type{T}) where {T<:BitSigned} = typeof(unsigned(zero(T)))

signed(::Type{T}) where {T<:Signed} = T
signed(::Type{T}) where {T<:BitUnsigned} = typeof(signed(zero(T)))
```
(I don't have v1.2 available to run these  .. if the foregoing does not work, there are bigger issues)

NHDaly commented:
> I am wondering what benefit there is to avoiding symmetry in these two functional definitions.

@JeffreySarnoff  I completely agree. We discussed it above, here: https://github.com/JuliaLang/julia/pull/30445#discussion_r243014703 -- the reason is to mirror the existing asymmetry in the definitions of `unsigned` and `signed`. As you pointed out there, I guess they're like that due to a difference regarding `BigInt`...

I opened https://github.com/JuliaLang/julia/issues/30473 to discuss this further. I think we should probably make `signed` and `unsigned` symmetric in the way you suggest. Either make them both take `Signed` and `Unsigned`, or both take `BitSigned` and `BitUnsigned`. I'd actually lean towards the first option i think. But yes, i agree, and i think we should change that for the existing `signed`/`unsigned` as well.

But that also seems orthogonal to this PR, which is why i separated it into #30473.

JeffreySarnoff commented:
I want to see this resolved as much as you do (modulo keypress as unit of effort).  There is nothing in my experience that suggests spackle over a conceptual inelegance is a winning approach -- there are situations wherein it is the only approach available, fortunately this is not so for us.

Rather than orthogonal concerns there is a dependance relationship.  Get the underpinning correct and this PR is implemented with clarity, coherence and aplomb.  Leaving the underpinning for resolution after the band-aid is applied ... well, who enjoys ripping off a bandage when the alternative is no bandage?

I don't mind that there are split into two PR.  It is my considered view that we need to come to an agreement on the other _first_.  I am willing to help do that.

JeffreySarnoff commented:
I want to see this resolved as much as you do (modulo keypress as unit of effort).  There is nothing in my experience that suggests spackle over a conceptual inelegance is a winning approach -- there are situations wherein it is the only approach available, fortunately this is not so for us.

Rather than orthogonal concerns there is a dependance relationship.  Get the underpinning correct and this PR is implemented with clarity, coherence and aplomb.  Leaving the underpinning for resolution after the band-aid is applied ... well, who enjoys ripping off a bandage when the alternative is no bandage?

I don't mind that there are split into two PR.  It is my considered view that we need to come to an agreement on the other _first_.  I am willing to help do that.

JeffreySarnoff commented:
I want to see this resolved as much as you do (modulo keypress as unit of effort).  There is nothing in my experience that suggests spackle over a conceptual inelegance is a winning approach -- there are situations wherein it is the only approach available, fortunately this is not so for us.

Rather than orthogonal concerns there is a dependance relationship.  Get the underpinning correct and this PR is implemented with clarity, coherence and aplomb.  Leaving the underpinning for resolution after the band-aid is applied ... well, who enjoys ripping off a bandage when the alternative is no bandage?

I don't mind that there are split into two PR.  It is my considered view that we need to come to an agreement on the other _first_.  I am willing to help do that.

JeffreySarnoff commented:
I have a feeling that's not all.  @timholy the first version of SaferIntegers was a conversion of your approach writing the pkg for Ints that round themselves -- and I put a premium on minimizing changes.  I remember spending days on why you had to use discrepant approaches to the sogned and the unsigned types and wondering how to bring both under a shared umbrella Integer type that would support the expressiveness we like.

JeffreySarnoff commented:
I have a feeling that's not all.  @timholy the first version of SaferIntegers was a conversion of your approach writing the pkg for Ints that round themselves -- and I put a premium on minimizing changes.  I remember spending days on why you had to use discrepant approaches to the sogned and the unsigned types and wondering how to bring both under a shared umbrella Integer type that would support the expressiveness we like.

NHDaly committed:
Merge branch 'master' into signed_T__unsigned_T

NHDaly committed:
Update compat note to Julia 1.4

NHDaly committed:
Restrict Base.[un]signed(::Type) to BitInteger by default.

Users can extend with their own types, just like for [un]signed(v).

NHDaly commented:
Now that #33459 was merged, I've (finally) updated this PR to the simple version suggested by @JeffreySarnoff above.
This is now fully symmetric, and simple. Thanks for the patience! :) 

@JeffBezanson on the review of #33459, you noted that you're "really surprised we still have `signed` and `unsigned`. It seems like those should be `x % Signed` instead."
Is it okay that we move forward then with extending these methods in this way? I suppose instead of overloading like this, we could simply rename these functions to something else, like `signedtype` and `unsignedtype` (or `signed_type` and `unsigned_type`)?

NHDaly commented:
> Is it okay that we move forward then with extending these methods in this way? I suppose instead of overloading like this, we could simply rename these functions to something else, like `signedtype` and `unsignedtype` (or `signed_type` and `unsigned_type`)?

Though note that, the whole motivation for this PR was the observation that Base already includes an (undocumented) method for `unsigned(::Type)`, just not one for `signed(::Type)`:
https://github.com/JuliaLang/julia/issues/30444#issue-392378387

(Which, actually, this makes me wonder why i didn't see a method redefinition warning.. Ah, never mind yeah there is one. Lemme fix that now.)

So I guess if we rename these, either we'd want to keep that weird undocumented asymmetry, or remove the undocumented method (which is "breaking")?

NHDaly committed:
Remove now unneeded extra def of `unsigned(::Type)` in multinverses.jl

NHDaly commented:
(Fixed the duplicate definition warning)

JeffreySarnoff commented:
fwiw I am happier with `Unsigned(T)` and `Signed(T)` than `unsigned(T)` and `signed(T)` for this role.

NHDaly commented:
> fwiw I am happier with `Unsigned(T)` and `Signed(T)` than `unsigned(T)` and `signed(T)` for this role.

I can see the appeal of that, but as a counter-argument, `Signed` and `Unsigned` read like _constructors for types_. It seems unexpected to me for the _Unsigned constructor_ to return a value that is not `<: Unsigned`.

Perhaps the best name for these would simply be the more explicit:
- `unsignedtype(T)` and `signedtype(T)`?

NHDaly commented:
I think that would be my vote now: `unsignedtype(T)` and `signedtype(T)`. It doesn't address the question of why we even have those `unsigned(x)` and `signed(x)` functions laying around, but oh well such is life. No reason to pile on to that decision.

`unsignedtype(T)` and `signedtype(T)` is clear, explicit, and discoverable, i think.

Keno commented:
This needs a rebase, but looks otherwise reasonable to me.

NHDaly commented:
@Keno do you have a preference regarding renaming these to `unsignedtype(T)` and `signedtype(T)`? If i was to rewrite this PR, i would advocate for those names (except that there's already the [undocumented] `unsigned(T)`, described in #30444)

Keno commented:
If it's just about @JeffBezanson's comment regarding the value of version these functions, then I don't think it's a problem, because the type method doesn't quite have the same semantics. It's not really a value conversion, it's a type property query. So I'm fine with the PR as is. Did I miss another reason to use a different generic function?

musm commented:
> unsignedtype(T) and signedtype(T) is clear, explicit, and discoverable, i think.

I also agree that these names would be more descriptive.  I'm not sure on the policy on renaming exported function in 1.x releases, but someone should chime in before we merge this.

kimikage commented:
BTW, PR #34864 was merged. So, at least #30444 could be closed.

The discussion in this PR may be still worth continuing, but it seems to be a good idea to change the subject of this PR.

cc: @NHDaly, @JeffreySarnoff

JeffreySarnoff commented:
what do you suggest?

kimikage commented:
To be honest, I am satisfied with the current `signed(T)`/`unsigned(T)`, so I don't feel the need to discuss new function names or interfaces.

However, I think that adding functions such as `unsignedtype` will affect PR #36526.

JeffreySarnoff commented:
That makes sense.  Certainly we should move the direction of these two so we are following one more consistent approach.

StefanKarpinski commented:
Marked for triage to decide whether to do this or not.

musm commented:
Sounds good, triage  should also be considered this PR and the methods names in light of  https://github.com/JuliaLang/julia/pull/36526 , as brought up by @kimikage and @JeffreySarnoff
